iceoryx_posh/doc/chunk_header.md

```diff
@@ -0,0 +1,323 @@
+
+The `Publisher` has additional template parameters for the custom header and payload alignment. Alternatively this could also be done with the `allocate` method, but that increases the risk of using it wrong by accident.
+
+Additionally, to fully integrate the custom header into the regular control flow, a pointer to a `ChunkHeaderHook` can be passed to the `Publisher` c'tor.
+
+The `ChunkHeaderHook` is not in the shared memory and has the following virtual methods:
+- `virtual void allocateHook(ChunkHeader& chunkHeader)` can be used to initialize the custom header
+- `virtual void deliveryHook(ChunkHeader& chunkHeader)` can be used to set members of the custom header, e.g. a timestamp
+
+Furthermore, the `Publisher` and `Subscriber` have access to the `ChunkHeader` and can use the `customHeader()` method to gain access to the custom header.
+
+### Pitfalls & Testing
+
+- when the payload is adjacent to the `ChunkHeader`, it must be ensured that `m_payloadOffset` overlaps with the `back-offset`, which is `sizeof(m_payloadOffset)` in front of the payload
+- to simplify calculation, it is assumed that the alignment of the custom header doesn't exceed the alignment of the `ChunkHeader`. This has to be enforced with an `assert`
+
+# Open Issues
+
+- the design was done with the intention to have a custom header of arbitrary size, if the size is limited to e.g. 32 bytes, some things could be simplified
+- typed publisher/subscriber API proposal
+```
+// publisher
+auto pub = iox::popo::TypedPublisher<MyPayload, MyHeader>(serviceDescription);
+pub.loan()
+    .and_then([&](auto& sample) {
+        sample.getHeader()->customHeader<MyHeader>().data = 42;
+        sample->a = 42;
+        sample->b = 13;
+        sample.publish();
+    })
+    .or_else([](iox::popo::AllocationError) {
+        // Do something with error.
+    });
+
+// subscriber
+auto sub = iox::popo::TypedSubscriber<MyPayload, MyHeader>(serviceDescription);
+sub->take()
+    .and_then([](auto& sample){
+        std::cout << "Custom header data: " << sample.getHeader()->customHeader<MyHeader>().data << std::endl;
+        std::cout << "Payload data: " << static_cast<const MyPayload*>(sample->get()).data << std::endl;
+    });
+```
+    - the publisher/subscriber would have a default parameter for the custom header to be source compatible with our current API
+    - the drawback is that the user could use the wrong custom header. Maybe `Sample` also needs an additional template parameter
+    - additionally, a `ChunkHeaderHook` could be used on the publisher side
+```
+template <typename CustomHeader>
+class MyChunkHeaderHook : public ChunkHeaderHook
+{
+  public:
+    void deliveryHook(ChunkHeader& chunkHeader) override
+    {
+        chunkHeader.customHeader<CustomHeader>().timestamp = myTimeProvider::now();
+    }
+};
+
+auto customHeaderHook = MyChunkHeaderHook<MyHeader>();
+auto pub = iox::popo::TypedPublisher<MyPayload>(serviceDescription, customHeaderHook);
```

Comment:
Don't you need MyHeader as template argument here? Or can you solve it with the provided Hook?

Reply:
No, that's the nice thing here. It's deduced from the Hook. Less things to get wrong :)